store.go

```diff
@@ -6,62 +6,133 @@ import (
 
 	log "github.com/Sirupsen/logrus"
 	"github.com/docker/libkv/store"
+	"github.com/docker/libnetwork/config"
 	"github.com/docker/libnetwork/datastore"
 )
 
-func (c *controller) validateDatastoreConfig() bool {
-	return c.cfg != nil && c.cfg.Datastore.Client.Provider != "" && c.cfg.Datastore.Client.Address != ""
+var (
+	defaultLocalStoreConfig = config.DatastoreCfg{
+		Embedded: true,
+		Client: config.DatastoreClientCfg{
+			Provider: "boltdb",
+			Address:  defaultPrefix + "/boltdb.db",
+			Config: &store.Config{
+				Bucket: "libnetwork",
+			},
+		},
+	}
+)
+
+func (c *controller) validateGlobalStoreConfig() bool {
+	return c.cfg != nil && c.cfg.GlobalStore.Client.Provider != "" && c.cfg.GlobalStore.Client.Address != ""
 }
 
-func (c *controller) initDataStore() error {
+func (c *controller) initGlobalStore() error {
 	c.Lock()
 	cfg := c.cfg
 	c.Unlock()
-	if !c.validateDatastoreConfig() {
-		return fmt.Errorf("datastore initialization requires a valid configuration")
+	if !c.validateGlobalStoreConfig() {
+		return fmt.Errorf("globalstore initialization requires a valid configuration")
 	}
 
-	store, err := datastore.NewDataStore(&cfg.Datastore)
+	store, err := datastore.NewDataStore(&cfg.GlobalStore)
 	if err != nil {
 		return err
 	}
 	c.Lock()
-	c.store = store
+	c.globalStore = store
 	c.Unlock()
+	return nil
+}
 
-	nws, err := c.getNetworksFromStore()
-	if err == nil {
-		c.processNetworkUpdate(nws, nil)
-	} else if err != datastore.ErrKeyNotFound {
-		log.Warnf("failed to read networks from datastore during init : %v", err)
+func (c *controller) initLocalStore() error {
+	c.Lock()
+	cfg := c.cfg
+	c.Unlock()
+	localStore, err := datastore.NewDataStore(c.getLocalStoreConfig(cfg))
+	if err != nil {
+		return err
+	}
+	c.Lock()
+	c.localStore = localStore
+	c.Unlock()
+	return nil
+}
+
+func (c *controller) processObjectsFromStore() error {
+	if c.globalStore != nil {
+		nws, err := c.getNetworksFromGlobalStore()
+		if err == nil {
+			c.processNetworkUpdate(nws, nil)
+		} else if err != datastore.ErrKeyNotFound {
+			log.Warnf("failed to read networks from globalstore during init : %v", err)
```

Comment:
Why are we not iterating endpoints from globalstore while we do it in localstore? In general for configuration both the stores should be treated exactly identically. In fact looking at this whole diff I am not a fan of having boiler plate code which does the same thing separately for globalstore and localstore. Once we decide and agree that the objects from global and local stores needs to be processed exactly the same way then we need to have common functions which operate on a list of stores and that's it. One of the property of the store could be locality which could be used for object placement purposes i.e another common function which WriteToSore but based on a KVObject Interface method called `Scope` will choose the right store to write the object into.